core/src/main/java/org/infinispan/commands/read/GetAllCommand.java

```diff
@@ -0,0 +1,204 @@
+import org.infinispan.container.InternalEntryFactory;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.context.Flag;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.distribution.ch.ConsistentHash;
+import org.infinispan.lifecycle.ComponentStatus;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.infinispan.commons.util.Util.toStr;
+
+/**
+ * Retrieves multiple entries at once.
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+// TODO: revise the command hierarchy, e.g. this should not implement MetadataAwareCommand
+public class GetAllCommand extends AbstractFlagAffectedCommand {
+   public static final byte COMMAND_ID = 44;
+   private static final Log log = LogFactory.getLog(GetAllCommand.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   private Collection<?> keys;
+   private boolean returnEntries;
+   private ConsistentHash ch;
+   private Address localAddress;
+
+   // TODO: remotely fetched are because of compatibility - can't we just always return InternalCacheEntry and have
+   //       the unboxing executed as the topmost interceptor?
+   private Map<Object, InternalCacheEntry> remotelyFetched;
+
+   private /* transient */ InternalEntryFactory entryFactory;
+
+   public GetAllCommand(Collection<?> keys, Set<Flag> flags,
+         boolean returnEntries, InternalEntryFactory entryFactory) {
+      this.keys = keys;
+      this.flags = flags;
+      this.returnEntries = returnEntries;
+      this.entryFactory = entryFactory;
+   }
+
+   GetAllCommand() {
+   }
+
+   @Override
+   public Object acceptVisitor(InvocationContext ctx, Visitor visitor) throws Throwable {
+      return visitor.visitGetAllCommand(ctx, this);
+   }
+
+   @Override
+   public boolean shouldInvoke(InvocationContext ctx) {
+      return true;
+   }
+
+   @Override
+   public boolean ignoreCommandOnStatus(ComponentStatus status) {
+      // TODO: this should only be ignored when it is ran in a remote context
+      return status != ComponentStatus.RUNNING && status != ComponentStatus.INITIALIZING;
+   }
+
+   @Override
+   public Object perform(InvocationContext ctx) throws Throwable {
+      Map<Object, Object> map = createMap();
+      for (Object key : keys) {
+         CacheEntry entry = ctx.lookupEntry(key);
+         if (entry == null) {
+            if (trace) {
+               log.tracef("Entry for key %s not found", key);
+            }
+            // We have to put null even if it isn't in the context.  This is because
+            // context won't have a value for null unless it is repeatable read.
+            if (ch == null || ch.isKeyLocalToNode(localAddress, key)) {
```

Comment:
Couldn't the distribution interceptor do this check and put the entry in the context before calling `invokeNext`?